flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sink/AbstractWriterOperator.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.InternalTimerService;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.util.List;
+
+/**
+ * Abstract base class for operators that work with a {@link Writer}.
+ *
+ * <p>Sub-classes are responsible for creating the specific {@link Writer} by implementing {@link
+ * #createWriter()}.
+ *
+ * @param <InputT> The input type of the {@link Writer}.
+ * @param <CommT> The committable type of the {@link Writer}.
+ */
+@Internal
+abstract class AbstractWriterOperator<InputT, CommT> extends AbstractStreamOperator<CommT>
+	implements OneInputStreamOperator<InputT, CommT>, BoundedOneInput {
+
+	private static final long serialVersionUID = 1L;
+
+	/** The runtime information of the input element. */
+	private final Context<InputT> context;
+
+	// ------------------------------- runtime fields ---------------------------------------
+
+	/** We listen to this ourselves because we don't have an {@link InternalTimerService}. */
+	private Long currentWatermark;
+
+	/** The sink writer that does most of the work. */
+	protected Writer<InputT, CommT, ?> writer;
+
+	AbstractWriterOperator() {
+		this.context = new Context<>();
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+
+		this.currentWatermark = Long.MIN_VALUE;
+
+		writer = createWriter();
+	}
+
+	@Override
+	public void processElement(StreamRecord<InputT> element) throws Exception {
+		context.element = element;
+		writer.write(element.getValue(), context);
+	}
+
+	@Override
+	public void prepareSnapshotPreBarrier(long checkpointId) throws Exception {
+		super.prepareSnapshotPreBarrier(checkpointId);
+		sendCommittables(writer.prepareCommit(false));
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		super.processWatermark(mark);
+		this.currentWatermark = mark.getTimestamp();
+	}
+
+	@Override
+	public void endInput() throws Exception {
+		sendCommittables(writer.prepareCommit(true));
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		writer.close();
+	}
+
+	protected Sink.InitContext createInitContext() {
+		return new Sink.InitContext() {
+			@Override
+			public int getSubtaskId() {
+				return getRuntimeContext().getIndexOfThisSubtask();
+			}
+
+			@Override
+			public MetricGroup metricGroup() {
+				return getMetricGroup();
+			}
+		};
+	}
+
+	/**
+	 * Creates and returns a {@link Writer}.
+	 *
+	 * @throws Exception If creating {@link Writer} fail
+	 */
+	abstract Writer<InputT, CommT, ?> createWriter() throws Exception;
+
+	private void sendCommittables(final List<CommT> committables) {
+		for (CommT committable : committables) {
+			output.collect(new StreamRecord<>(committable));
```

Comment:
for the Iceberg sink case where the GlobalCommitter runs on jobmanager, this will continue to work, right?

Reply:
For the first version we plan to put the `GlobaCommitter`  in a normal operator. I would open a jira for following track.